### DISCUSSION TITLE
pyraf cannot find login.cl

### DISCUSSION THREAD

**Discussion by ljtpanda** [iraf-community/iraf#383] [Q&A] [answered]
Hello, Dear All,

I'm quite new to iraf and pyraf. I just installed them on my MacOS. When running pyraf, I got the error message:
```
Warning: no login.cl found
```
This may further cause some other errors when running other commands, such as:
```
--> imstat dev$pix
Traceback (innermost last):
  File "<CL script CL1>", line 1, in <module>
pyraf.tools.irafglobals.IrafError: Undefined variable `uparm' in string `uparm$/imlimstas.par'
```
I tried to create a symbolic of login.cl to the working directory, or run mkiraf -c to create it. Both work well and the error messages disappear. However, I don't want to create login.cl every time I run pyraf in a new directory. It seems if I don't do this, pyraf cannot find the correct login.cl in my iraf directory automatically. How to solve the problem?

Thank you very much.

**Comment by olebole** [comment]
Please always specify which version of IRAF, PyRAF and Python you are using, and how you installed it.

PyRAF (2.2.1) searches `login.cl` on a number of places:

1. in the current directory,
2.  `~/.iraf/login.cl`,
3. `/etc/iraf/login.cl`,
4. `hlib$login.cl`

with `hlib$` usually set to `iraf$unix/hlib`if not defined as environment variable. The last one is the final resort and should always work. Can you give the result of 

```
--> osfn("hlib$")
```

on your PyRAF prompt (should be `/usr/local/lib/iraf/unix/hlib/` or similarly pointing to your IRAF installation), and check that there is a `login.cl` in this directory?

**Reply by ljtpanda** [reply]
Thank you very much.

I installed IRAF 2.17.1 and pyraf 2.2.1. I think the python I used is that installed via anaconda, which is 3.11.7.

For IRAF, I installed the source code. For pyraf, I installed it using pip3 install pyraf.

Here is the result checking hlib$:
--> osfn("hlib$")
'/usr/local/lib/iraf/unix/hlib/'

There is a login.cl in this direction. And if I created a link of it in the working directory:
ln -s /usr/local/lib/iraf/unix/hlib/login.cl login.cl
I can start pyraf with no warnings. So I think the file is correct.

BTW, I also found another two login.cl files:

find /usr/local/lib/iraf -name login.cl 

/usr/local/lib/iraf/unix/hlib/login.cl
/usr/local/lib/iraf/pkg/cl/login.cl
/usr/local/lib/iraf/pkg/ecl/login.cl

I'm not sure why pyraf cannot find login.cl in hlib$

**Reply by olebole** [reply, marked as answer]
OK, I found the cause. The search in `hlib$` was added only after version 2.2.1 (in https://github.com/iraf-community/pyraf/pull/150). You should therefore create a user default one with `mkiraf -c -d` for the time being. This will be fixed with the next version.

**Reply by ljtpanda** [reply]
Thank you very much. This solves the problem.
